acados/sim/sim_gnsf.c

```diff
@@ -0,0 +1,1757 @@
+
+// blasfeo
+#include "blasfeo/include/blasfeo_target.h"
+#include "blasfeo/include/blasfeo_common.h"
+#include "blasfeo/include/blasfeo_d_aux_ext_dep.h"
+#include "blasfeo/include/blasfeo_d_blas.h"
+#include "blasfeo/include/blasfeo_d_kernel.h"
+#include "blasfeo/include/blasfeo_i_aux_ext_dep.h"
+#include "blasfeo/include/blasfeo_d_aux.h"
+
+
+/************************************************
+* dims
+************************************************/
+
+int sim_gnsf_dims_calculate_size()
+{
+    int size = sizeof(sim_gnsf_dims);
+    return size;
+}
+
+
+void *sim_gnsf_dims_assign(void* config_, void *raw_memory)
+{
+    char *c_ptr = (char *) raw_memory;
+    sim_gnsf_dims *dims = (sim_gnsf_dims *) c_ptr;
+    c_ptr += sizeof(sim_gnsf_dims);
+    assert((char *) raw_memory + sim_gnsf_dims_calculate_size() == c_ptr);
+    return dims;
+}
+
+/************************************************
+* get & set functions
+************************************************/
+
+void sim_gnsf_set_nx(void *dims_, int nx)
+{
+    sim_gnsf_dims *dims = (sim_gnsf_dims *) dims_;
+    dims->nx = nx;
+}
+
+void sim_gnsf_set_nu(void *dims_, int nu)
+{
+    sim_gnsf_dims *dims = (sim_gnsf_dims *) dims_;
+    dims->nu = nu;
+}
+
+void sim_gnsf_get_nx(void *dims_, int* nx)
+{
+    sim_gnsf_dims *dims = (sim_gnsf_dims *) dims_;
+    *nx = dims->nx;
+}
+
+void sim_gnsf_get_nu(void *dims_, int* nu)
+{
+    sim_gnsf_dims *dims = (sim_gnsf_dims *) dims_;
+    *nu = dims->nu;
+}
+
+/************************************************
+* import functions
+************************************************/
+
+void gnsf_get_dims( sim_gnsf_dims *dims, casadi_function_t get_ints_fun)
+{
+    double *ints_out;
+    ints_out = (double*) calloc(10,sizeof(double));
+    export_from_ML_wrapped(ints_out, ints_out, get_ints_fun);
+
+    dims->nx        = (int) ints_out[0];
+    dims->nu        = (int) ints_out[1]; 
+    dims->nz        = (int) ints_out[2];
+    dims->nx1       = (int) ints_out[3];
+    dims->nx2       = (int) ints_out[4];
+    dims->n_out     = (int) ints_out[5];
+    dims->ny        = (int) ints_out[6];
+    dims->nuhat     = (int) ints_out[7];
+    dims->num_stages= (int) ints_out[8];
+    dims->num_steps = (int) ints_out[9];
+
+    free(ints_out);
+}
+
+
+void gnsf_import_matrices(sim_gnsf_dims* dims, gnsf_model *model, external_function_generic *get_matrices_fun)
+{
+    int nu  = dims->nu;
+    int nx1 = dims->nx1;
+    int nx2 = dims->nx2;
+    int nz = dims->nz;
+    int n_out = dims->n_out;
+    int ny = dims->ny;
+    int nuhat = dims->nuhat;
+    int exported_doubles = 0;
+
+    exported_doubles += (nx1 + nz) * (nx1 + nu + n_out + nx1+nz); // A, B, C, E;
+    exported_doubles += ny * (2*nx1 + nz); // L_x, L_xdot, L_z, L_u;
+    exported_doubles += nuhat * nu; // L_u;
+    exported_doubles += nx2*nx2; // A_LO;
+    // printf("exported_doubles %d= \n",exported_doubles);
+    double *export_in  = (double*) malloc(1*sizeof(double));
```

Comment:
there should be no malloc and free in acados core

workspace and memory are here for this